### PR TITLE
Castle

### PR DESCRIPTION
**Pull request by dtrypt**
**UPDATE: Rebased to the latest and tests pass finally.**
- Adding castling that follows this guide:
http://www.mark-weeks.com/aboutcom/ble14cas.htm
- Added a 'castle' field to pieces model which is for checking is the kings/rooks have been moved prior to castling.
- Moved most logic from piece model to pieces_helper.
- Refactored move_to! & capture logic.
- Had to use a lot of methods to appease rubocops complexity checks, which made it even more complex imo =/.
- Added an alert pop-up to the view for invalid moves.
- Added tests for castling and my backlogged tests for updating pieces.

- Please review this PR before PRs that come after this, or there will be tons of merge conflicts because I changed A LOT of files.
- I need to add "prevent yourself from moving into check" to wrap up the last 2 conditions. I'm going to submit an additional PR for that after we merge in Leslie's part.

### CODE REVIEW

File: app/helpers/pieces_helper.rb [hunk 1]
```diff
@@ -1,2 +1,99 @@
 module PiecesHelper
```
Comment (sungwon) [inline app/helpers/pieces_helper.rb:1]: Good idea refactoring this out into a Helper

File: app/models/game.rb [hunk 2]
```diff
@@ -23,7 +23,7 @@ def populate_board!
   end
 
   def check?(color)
-    king = King.where(game: self, color: color)
+    king = King.find_by(game: self, color: color)
```
Comment (sungwon) [inline app/models/game.rb:26]: Nice. This is neater than .where(....).first

File: app/models/king.rb [hunk 3]
```diff
@@ -4,4 +4,48 @@ def valid_move?(x_end, y_end)
...
+  def castle_check(new_x, color)
+    define_rooks
+    # Need "prevent yourself from moving into check" method to cover the remaining restrictions
+    return true if new_x == 3 && queen_side_castle_check(color)
+    return true if new_x == 7 && king_side_castle_check(color)
+    false
+  end
+
+  def queen_side_castle_check(color)
+    rook = instance_variable_get("@#{color}_queen_rook")
```
Comment (sungwon) [inline app/models/king.rb]: Whoa! Nice meta-programming. 

How about using this technique to reduce the code duplication and combine queen_side_castle_check and king_side_castle_check into, say: 

```ruby
def castle_check(color, side)
  rook = instance_variable_get("@#{color}_#{side}_rook")
...
```

File: app/models/king.rb [hunk 4]
```diff
@@ -4,4 +4,48 @@ def valid_move?(x_end, y_end)
...
+  end
+
+  def perform_castling(new_x, color)
+    define_rooks
+    update_attributes(horizontal_position: new_x, castle: false)
+    return perform_queenside_castling(color) if new_x == 3
+    return perform_kingside_castling(color) if new_x == 7
+  end
+
+  def perform_queenside_castling(color)
```
Comment (sungwon) [inline app/models/king.rb:29]: Combine with kingside into one perform_castling method just like castle_check

File: app/views/games/show.html.erb [hunk 5]
```diff
@@ -16,15 +16,18 @@
...
-        });
+        $(ui.draggable).detach().css({top: 0, left: 0}).appendTo(this);
+          $.ajax({
+            type: 'PUT',
+            url: desiredUrl,
+            data: piece,
+            dataType: 'json'
+          }).done(function(data){
+            if (data.horizontal_position !== x || data.vertical_position !== y) {
+              alert("Illegal move. Try again.");
```
Comment (sungwon) [inline app/views/games/show.html.erb:27]: Good idea.

File: app/helpers/pieces_helper.rb [hunk 6]
```diff
@@ -1,2 +1,99 @@
...
+  def can_capture?(new_x, new_y)
```
Comment (sungwon) [inline app/helpers/pieces_helper.rb]: I think we should rename this to just `capture`.  `can_capture?` sounds like it just returns a boolean if it is able to capture, but it has the unexpected side effect of actually performing the capture and updating the piece position. If we say `capture` we know we are trying to capture, but simply return false if we can't.

**Review by sungwon — CHANGES_REQUESTED** [review]
Epic pull request! Nice work

**Review by koglesby — COMMENTED** [review]
Awesome castling logic - works well, and nice refactoring. I'm not sure why the chess#index tests are failing though

**Comment by dtrypt** [comment]
To my surprise I was getting the issue with the same 2 tests that were failing even after rebasing with the latest merge. Data from past tests were persisting into new tests. It seems like this is a common issue and some digging pointed me to "database_cleaner" - a ruby gem that allows you to reset the database before running tests. Seems like this is the best solution and everything works properly.

**Review by sungwon — APPROVED** [review]

**Comment by sungwon** [comment]
Good find with that database_cleaner. That's weird, I was under the impression that the test database was cleared automatically after every test session. Guess not!

**Review by koglesby — APPROVED** [review]

**Pull request merged**